tools/traffic/workers/blob_reader.go

```diff
@@ -0,0 +1,231 @@
+	operatorSuccessMetrics     map[core.OperatorID]metrics.CountMetric
+	operatorFailureMetrics     map[core.OperatorID]metrics.CountMetric
+	requiredReadPoolSizeMetric metrics.GaugeMetric
+	optionalReadPoolSizeMetric metrics.GaugeMetric
+}
+
+// NewBlobReader creates a new BlobReader instance.
+func NewBlobReader(
+	ctx *context.Context,
+	waitGroup *sync.WaitGroup,
+	logger logging.Logger,
+	config *config.WorkerConfig,
+	retriever clients.RetrievalClient,
+	chainClient eth.ChainClient,
+	blobStore *table.BlobStore,
+	generatorMetrics metrics.Metrics) BlobReader {
+
+	return BlobReader{
+		ctx:         ctx,
+		waitGroup:   waitGroup,
+		logger:      logger,
+		config:      config,
+		retriever:   retriever,
+		chainClient: chainClient,
+		blobsToRead: blobStore,
+		metrics: &blobReaderMetrics{
+			generatorMetrics:           generatorMetrics,
+			fetchBatchHeaderMetric:     generatorMetrics.NewLatencyMetric("fetch_batch_header"),
+			fetchBatchHeaderSuccess:    generatorMetrics.NewCountMetric("fetch_batch_header_success"),
+			fetchBatchHeaderFailure:    generatorMetrics.NewCountMetric("fetch_batch_header_failure"),
+			recombinationSuccessMetric: generatorMetrics.NewCountMetric("recombination_success"),
+			recombinationFailureMetric: generatorMetrics.NewCountMetric("recombination_failure"),
+			readLatencyMetric:          generatorMetrics.NewLatencyMetric("read"),
+			validBlobMetric:            generatorMetrics.NewCountMetric("valid_blob"),
+			invalidBlobMetric:          generatorMetrics.NewCountMetric("invalid_blob"),
+			readSuccessMetric:          generatorMetrics.NewCountMetric("read_success"),
+			readFailureMetric:          generatorMetrics.NewCountMetric("read_failure"),
+			operatorSuccessMetrics:     make(map[core.OperatorID]metrics.CountMetric),
+			operatorFailureMetrics:     make(map[core.OperatorID]metrics.CountMetric),
+			requiredReadPoolSizeMetric: generatorMetrics.NewGaugeMetric("required_read_pool_size"),
+			optionalReadPoolSizeMetric: generatorMetrics.NewGaugeMetric("optional_read_pool_size"),
+		},
+	}
+}
+
+// Start begins a blob reader goroutine.
+func (r *BlobReader) Start() {
+	r.waitGroup.Add(1)
+	ticker := time.NewTicker(r.config.ReadRequestInterval)
+	go func() {
+		defer r.waitGroup.Done()
+		for {
+			select {
+			case <-(*r.ctx).Done():
+				err := (*r.ctx).Err()
+				if err != nil {
+					r.logger.Info("blob r context closed", "err:", err)
+				}
+				return
+			case <-ticker.C:
+				r.randomRead()
+			}
+		}
+	}()
+}
+
+// randomRead reads a random blob.
+func (r *BlobReader) randomRead() {
+	metadata := r.blobsToRead.GetNext()
+	if metadata == nil {
+		// There are no blobs that we are required to read.
+		return
+	}
+
+	r.metrics.requiredReadPoolSizeMetric.Set(float64(r.blobsToRead.Size()))
+
+	ctxTimeout, cancel := context.WithTimeout(*r.ctx, r.config.FetchBatchHeaderTimeout)
+	defer cancel()
+	batchHeader, err := metrics.InvokeAndReportLatency(r.metrics.fetchBatchHeaderMetric,
```

Comment:
I wouldn't consider a mix of core logic with monitoring code quite clean, and would prefer a separation.

Reply:
There are a few patterns I could utilize for reporting latency. Which do you prefer?

# Pattern A

Wrap the base function in another function that calls the base function and records the amount of time it took to execute. (This is the pattern currently in the PR)

# Pattern B

Measure the time before and after the function call.

```
start := time.Now()
foo()
end := time.Now()
metrics.reportDurationOfFoo(start, end)
```

# Pattern C

Embed the measurement of the time to execute the method inside the method itself.

```
func foo() {
    start := time.now()
    // business logic
    end := time.now()
    metrics.reportDurationOfFoo(start, end)
```

# Pattern D

Is there a strategy you like that I didn't mention?

Reply:
I've removed the `invokeAndReportLatency()` metric in favor of pattern B